fl4health/clients/scaffold_client.py

```diff
@@ -69,24 +68,34 @@ def get_parameters(self, config: Config) -> NDArrays:
 
     def set_parameters(self, parameters: NDArrays, config: Config) -> None:
         """
-        Assumes that the parameters being passed contain model parameters concatenated with
-        server control variates. They are unpacked for the clients to use in training
+        Assumes that the parameters being passed contain model parameters concatenated with server control variates.
+        They are unpacked for the clients to use in training. If it's the first time the model is being initialized,
+        we assume the full model is being initialized and use the FullParameterExchanger() to set all model weights
+        Args:
+            parameters (NDArrays): Parameters have information about model state to be added to the relevant client
+                model and also the server control variates (initial or after aggregation)
+            config (Config): The config is sent by the FL server to allow for customization in the function if desired.
         """
         assert self.model is not None and self.parameter_exchanger is not None
 
         server_model_state, server_control_variates = self.parameter_exchanger.unpack_parameters(parameters)
         self.server_control_variates = server_control_variates
-        self.server_model_state = server_model_state
-        self.parameter_exchanger.pull_parameters(server_model_state, self.model, config)
+
+        super().set_parameters(server_model_state, config)
+
+        # Note that these are weights the require a gradient, because they are used to compute control variates
         self.server_model_weights = [
-            model_params.cpu().detach().numpy()
+            model_params.cpu().detach().clone().numpy()
             for model_params in self.model.parameters()
             if model_params.requires_grad
         ]
 
-        # If client control variates do not exist, initialize with zeros as per paper
+        # If client control variates do not exist, initialize them to be the same as the server control variates.
+        # Server variates default to be 0, but as stated in the paper the control variates should be the uniform
+        # average of the client variates. So if server_control_variates are non-zero, this ensures that average
+        # still holds.
         if self.client_control_variates is None:
-            self.client_control_variates = [np.zeros_like(weight) for weight in self.server_control_variates]
+            self.client_control_variates = copy.deepcopy(self.server_control_variates)
```

Comment:
This used to be setting it to zeros, which if the user explicitly sets the initial server_control_variates to be something else, say all ones, would violate the assumption of the paper that the initial values for self.server_control_variates = average(client_control_variates). So here we're initializing the control variates to match the server ones

Reply:
Nice that we can correctly handle init of control variates in all cases. Thanks for the update :)